### ISSUE TITLE
Set up and load data onto cluster image

### ISSUE THREAD

**Issue by LeopoldEhrlich** [LeopoldEhrlich/EEG-SSL#3]

**Comment by Sebastian-Medrea** [comment]
Will do.

**Comment by Sebastian-Medrea** [comment]
Not too sure if this is what we are looking for, but I wrote up a script for the 5 BENDR datasets to extract relevant band features and ERP for labelling, in addition to a PCA and histogram visualization for exploratory data analysis. Tested this for MMI atm only

`'''
General EEG setup and loader as a cluster image and feature extractor/visualization for BENDR paper datasets.

BENDR Benchmark Datasets:

MMI (Motor Movement/Imagery): motor_imagery, alpha/beta
BCIC (BCI Competition IV 2a): motor_imagery, alpha/beta
ERN (Error-Related Negativity): erp, P300/N400
P300 (ERP-BCI speller): erp, P300/N400
SSC (Sleep Stage Classification): sleep, delta/theta/alpha/beta

Usage:

For MMI/BCIC: dataset_type='motor_imagery' (frequency bands)
For ERN/P300: dataset_type='erp' (event-related potentials)
For SSC: dataset_type='sleep' (all frequency bands)
'''
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path
import pandas as pd

import mne
from scipy import signal
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

mne.set_log_level('WARNING')
# Dataset type configurations
DATASET_CONFIGS = {
    'motor_imagery': {
        'features': ['alpha', 'beta', 'power'],
        'description': 'Motor imagery (MMI, BCIC)',
        'epoch_length': 2.0
    },
    'erp': {
        'features': ['p300', 'n400', 'power'],
        'description': 'ERP tasks (P300, ERN)',
        'epoch_length': 1.0
    },
    'sleep': {
        'features': ['delta', 'theta', 'alpha', 'beta'],
        'description': 'Sleep staging (SSC)',
        'epoch_length': 30.0
    }
}


class GeneralEEGFeatureExtractor:
    '''
    Universal feature extractor for different EEG paradigms.
    Auto-detects dataset type or uses specified type.
    '''
    def __init__(self, sfreq=160, dataset_type='auto'):

        self.sfreq = sfreq #sampling frequency set to 160 Hz as an example (MMI uses 160, BCIC uses 250)
        self.dataset_type = dataset_type
    
    def get_band_power(self, epochs, freq_range):
        '''Extract power in a specific frequency band.'''

        # Butterworth filter of order 4 grabs frequencies between freq_range[0] and freq_range[1] (a higher order is sharper but prone to noise)
        data = epochs.get_data()
        sos = signal.butter(4, freq_range, btype='bandpass', fs=self.sfreq, output='sos')
        
        # Apply each filter to each epoch and calculate the power and average out aross all time points and channels
        powers = []
        for epoch in data:
            filtered = signal.sosfilt(sos, epoch, axis=1)
            power = np.mean(filtered ** 2)
            powers.append(power)
        
        return np.array(powers)
    
    # Grabbing each frequency band
    def get_delta_power(self, epochs):
        '''Delta (0.5-4 Hz)'''

        return self.get_band_power(epochs, [0.5, 4])
    
    def get_theta_power(self, epochs):
        '''Theta (4-8 Hz)'''

        return self.get_band_power(epochs, [4, 8])
    
    def get_alpha_power(self, epochs):
        '''Alpha (8-13 Hz)'''

        return self.get_band_power(epochs, [8, 13])
    
    def get_beta_power(self, epochs):
        '''Beta (13-30 Hz): active thinking, motor activity'''

        return self.get_band_power(epochs, [13, 30])
    
    # Grabbing ERP features for labelling
    def get_p300_erp(self, epochs):
        '''
        P300: positive peak ~300ms after stimulus.

        Pz is the parietal midline and Cz is the central midline where P300 is strongest; the time window is 250-400ms after stimulus
        since our peak is around 300 ms. We find the maximum value or voltage spike in that window.
        '''
        
        # Look at specific channels, look at specific time window, and find the maximum amplitude in that window
        try:
            data = epochs.copy().pick_channels(['Pz', 'Cz'], ordered=True).get_data()

        except:
            data = epochs.get_data()[:, :2, :]
        
        times = epochs.times
        window = np.where((times >= 0.25) & (times <= 0.4))[0]

        if len(window) == 0:
            return np.zeros(len(epochs))

        return np.max(np.mean(data[:, :, window], axis=1), axis=1)
    
    def get_n400_erp(self, epochs):
        '''
        N400: negative dip ~400ms after semantic violation
        
        Fz is the frontal midline and Cz is the central midline where N400 is strongest; the time window is 350-500ms since our dip is around 400 ms. We find the minimum value or voltage dip in that window.
        '''

        try:
            data = epochs.copy().pick_channels(['Fz', 'Cz'], ordered=True).get_data()

        except:
            data = epochs.get_data()[:, :2, :]
        
        times = epochs.times
        window = np.where((times >= 0.35) & (times <= 0.5))[0]

        if len(window) == 0:
            return np.zeros(len(epochs))

        return np.min(np.mean(data[:, :, window], axis=1), axis=1)
    
    def get_mean_power(self, epochs):
        '''Total power across all frequencies.'''

        data = epochs.get_data()
        return np.mean(data ** 2, axis=(1, 2))
    
    def detect_dataset_type(self, epochs):
        '''
        Auto-detect the EEG data typebased on epoch characteristics.

        Sleep data tends to have longer 30s epochs whereas ERP data is short (1s) and motor imagery data is in between.
        '''

        epoch_length = epochs.times[-1] - epochs.times[0]
        
        if epoch_length > 20:
            return 'sleep'

        elif epoch_length < 1.5:
            return 'erp'

        else:
            return 'motor_imagery'
    
    def extract_features(self, epochs):
        '''Extract features based on dataset type.'''

        # determine type
        if self.dataset_type == 'auto':
            actual_type = self.detect_dataset_type(epochs)
            print(f"  Auto-detected: {actual_type}")

        else:
            actual_type = self.dataset_type
        
        # extract appropriate features
        if actual_type == 'motor_imagery':
            features = {
                'alpha': self.get_alpha_power(epochs),
                'beta': self.get_beta_power(epochs),
                'power': self.get_mean_power(epochs)
            }

            feature_names = ['alpha', 'beta', 'power']
            
        elif actual_type == 'erp':
            features = {
                'p300': self.get_p300_erp(epochs),
                'n400': self.get_n400_erp(epochs),
                'power': self.get_mean_power(epochs)
            }

            feature_names = ['p300', 'n400', 'power']
            
        elif actual_type == 'sleep':
            features = {
                'delta': self.get_delta_power(epochs),
                'theta': self.get_theta_power(epochs),
                'alpha': self.get_alpha_power(epochs),
                'beta': self.get_beta_power(epochs)
            }

            feature_names = ['delta', 'theta', 'alpha', 'beta']
        
        else:
            raise ValueError(f"Unknown dataset type: {actual_type}")
        
        # Normalize for clustering for 0 mean and 1 std so bands are on the same scale
        scaler = StandardScaler()
        combined = np.column_stack([features[f] for f in feature_names])
        combined_norm = scaler.fit_transform(combined)
        
        features['combined'] = combined_norm
        features['feature_names'] = feature_names
        features['type'] = actual_type
        
        return features
    
    def cluster_data(self, features, n_clusters=3):
        '''
        Cluster features into brain states. Using 3 as an example, cluster 0 would be baseline/resting, cluster 1 would be active motor imagery,
        and cluster 2 could be a transitional state.
        '''

        kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init=10)

        return kmeans.fit_predict(features)


def process_file(file_path, n_clusters=3, epoch_length=None, dataset_type='auto'):
    '''
    Process a single EEG file for testing. Loads raw EEG, creates fixed-length epochs, extracts features, and clusters.
    
    Args:
        file_path: path to .edf or .fif file
        n_clusters: number of clusters to find
        epoch_length: epoch duration (None = auto-select based on type)
        dataset_type: 'motor_imagery', 'erp', 'sleep', or 'auto'
    '''

    file_path = str(file_path)
    
    # load data in proper extension
    if file_path.endswith('.edf'):
        raw = mne.io.read_raw_edf(file_path, preload=True)

    elif file_path.endswith('.fif'):
        raw = mne.io.read_raw_fif(file_path, preload=True)

    else:
        raise ValueError("Need .edf or .fif file")
    
    # auto-select epoch length if not specified
    if epoch_length is None:

        if dataset_type in DATASET_CONFIGS:
            epoch_length = DATASET_CONFIGS[dataset_type]['epoch_length']
        else:
            epoch_length = 2.0  # default
    
    # make epochs
    events = mne.make_fixed_length_events(raw, duration=epoch_length)
    epochs = mne.Epochs(raw, events, tmin=0, tmax=epoch_length,
                       baseline=None, preload=True, verbose=False)
    
    # extract features
    extractor = GeneralEEGFeatureExtractor(sfreq=raw.info['sfreq'], dataset_type=dataset_type)
    features = extractor.extract_features(epochs)
    
    # cluster
    labels = extractor.cluster_data(features['combined'], n_clusters=n_clusters)
    
    return {
        'file': file_path,
        'n_epochs': len(epochs),
        'features': features,
        'labels': labels,
        'sfreq': raw.info['sfreq'],
        'n_channels': len(raw.ch_names),
        'duration': raw.times[-1],
        'dataset_type': features['type']
    }


def batch_process_files(file_list, n_clusters=3, epoch_length=None, output_dir='processed_data', dataset_type='auto'):
    '''
    Process multiple files and save features. Processes each file, saves individual CSVs as feature summary and a combined CSV for all files.
    
    Args:
        file_list: list of file paths
        n_clusters: number of clusters
        epoch_length: epoch duration (None = auto)
        output_dir: where to save results
        dataset_type: 'motor_imagery', 'erp', 'sleep', or 'auto'
    '''

    output_dir = Path(output_dir)
    output_dir.mkdir(exist_ok=True, parents=True)
    
    print(f"Processing {len(file_list)} files")

    if dataset_type != 'auto':
        print(f"Dataset type: {DATASET_CONFIGS[dataset_type]['description']}")

    print(f"Output: {output_dir}")

    
    all_results = []
    all_features = []
    combined_df = None
    
    for i, file_path in enumerate(file_list, 1):
        print(f"\n[{i}/{len(file_list)}] {Path(file_path).name}")
        
        try:
            result = process_file(file_path, n_clusters, epoch_length, dataset_type)
            
            # print summary
            print(f"Type: {result['dataset_type']}")
            print(f"Duration: {result['duration']:.1f}s, Epochs: {result['n_epochs']}")
            
            # print feature ranges
            for feat_name in result['features']['feature_names'][:2]:
                feat_data = result['features'][feat_name]
                print(f"  {feat_name}: [{feat_data.min():.3e}, {feat_data.max():.3e}]")
            
            print(f"Clusters: {np.bincount(result['labels'])}")
            
            # save individual file
            file_name = Path(file_path).stem
            df_data = {'epoch': range(result['n_epochs']), 'cluster': result['labels']}

            for feat_name in result['features']['feature_names']:
                df_data[feat_name] = result['features'][feat_name]
            
            df = pd.DataFrame(df_data)
            df.to_csv(output_dir / f'{file_name}_features.csv', index=False)
            
            all_results.append(result)
            all_features.append(df)
            
        except Exception as e:

            print(f"  ERROR: {e}")
            continue
    
    # combine all
    if all_features:

        combined_df = pd.concat(all_features, ignore_index=True)
        combined_df.to_csv(output_dir / 'all_features_combined.csv', index=False)
        
        print(f"Combined: {output_dir / 'all_features_combined.csv'}")
        print(f"Total epochs: {len(combined_df)}")
        
    else:
        print("\nNo files processed successfully!")
    
    return all_results, combined_df


def visualize_features(features_csv, show_plot=True):
    '''Visualize extracted features (adapts to feature types)'''
    df = pd.read_csv(features_csv)
    
    print("\nFeature Statistics:")
    print(df.describe())
    
    print("\nCluster Distribution:")
    print(df['cluster'].value_counts().sort_index())
    
    if show_plot:

        feature_cols = [col for col in df.columns if col not in ['epoch', 'cluster']]
        
        fig, axes = plt.subplots(2, 2, figsize=(12, 10))
        
        # Plot first two features
        if len(feature_cols) >= 1:
            axes[0, 0].hist(df[feature_cols[0]], bins=30, edgecolor='black', alpha=0.7)
            axes[0, 0].set_xlabel(feature_cols[0].capitalize())
            axes[0, 0].set_title(f'{feature_cols[0].capitalize()} Distribution')
            axes[0, 0].grid(True, alpha=0.3)
        
        if len(feature_cols) >= 2:
            axes[0, 1].hist(df[feature_cols[1]], bins=30, edgecolor='black', alpha=0.7, color='orange')
            axes[0, 1].set_xlabel(feature_cols[1].capitalize())
            axes[0, 1].set_title(f'{feature_cols[1].capitalize()} Distribution')
            axes[0, 1].grid(True, alpha=0.3)
        
        # Cluster sizes
        cluster_counts = df['cluster'].value_counts().sort_index()
        axes[1, 0].bar(cluster_counts.index, cluster_counts.values, edgecolor='black')
        axes[1, 0].set_xlabel('Cluster ID')
        axes[1, 0].set_title('Cluster Sizes')
        axes[1, 0].grid(True, alpha=0.3)
        
        # 2D cluster plot
        features_array = df[feature_cols].values
        scaler = StandardScaler()
        features_norm = scaler.fit_transform(features_array)
        
        pca = PCA(n_components=2)
        features_2d = pca.fit_transform(features_norm)
        
        scatter = axes[1, 1].scatter(features_2d[:, 0], features_2d[:, 1],
                                    c=df['cluster'], cmap='tab10',
                                    alpha=0.6, edgecolors='black', linewidth=0.5)
        axes[1, 1].set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]:.1%})')
        axes[1, 1].set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]:.1%})')
        axes[1, 1].set_title('2D Cluster Visualization')
        axes[1, 1].grid(True, alpha=0.3)
        plt.colorbar(scatter, ax=axes[1, 1], label='Cluster')
        
        plt.tight_layout()
        
        output_path = Path(features_csv).parent / 'feature_visualization.png'
        plt.savefig(output_path, dpi=150, bbox_inches='tight')
        print(f"\nVisualization saved: {output_path}")
        
        plt.show()
    
    return df


def main():
    '''
    Process BENDR benchmark datasets:
    
    Tested on MMI
    
    When running the others,

    BCIC: dataset_type='motor_imagery'
    ERN: dataset_type='erp'
    P300: dataset_type='erp'
    SSC: dataset_type='sleep'
    
    Or use 'auto' to detect automatically.
    '''
    
    data_dir = Path('data/mmidb')
    
    if not data_dir.exists():
        print("Data directory not found!")
        print("Put your .edf files in: data/mmidb/")

    else:

        eeg_files = list(data_dir.rglob('*.edf'))
        
        if not eeg_files:
            print("No .edf files found!")

        else:
            print(f"Found {len(eeg_files)} EEG files\n")
            
            
            results, combined_df = batch_process_files(
                file_list=eeg_files,
                n_clusters=3,
                epoch_length=None,  # auto-select based on type
                output_dir='processed_data',
                dataset_type='motor_imagery'  # Edit based on type
            )
            
            # Visualize
            if combined_df is not None:
                
                print("Generating visualization:")
                visualize_features('processed_data/all_features_combined.csv', show_plot=True)
                
                print("DONE!")
                print(f"Processed {len(results)} files")
                print("Features saved to: processed_data/")

main()

[cluster_image_loader.py](https://github.com/user-attachments/files/24749807/cluster_image_loader.py)

`
